src/sentry/snuba/discover.py

```diff
@@ -946,57 +948,80 @@ def measurements_histogram_query(
+        measurements = []
+        for f in fields:
+            measurement = get_measurement_name(f)
+            if measurement is None:
+                raise InvalidSearchQuery(
+                    u"multihistogram expected all measurements, received: {}".format(f)
+                )
+            measurements.append(measurement)
+        conditions.append([key_alias, "IN", measurements])
 
-    conditions = [[get_function_alias(key_col), "IN", measurements]]
+    histogram_params = find_histogram_params(num_buckets, min_value, max_value, multiplier)
+    histogram_column = get_histogram_column(fields, key_column, histogram_params)
+    histogram_alias = get_function_alias(histogram_column)
 
     # make sure to bound the bins as to not get too many results
     if min_value is not None:
         min_bin = histogram_params.start_offset
-        conditions.append([get_function_alias(histogram_col), ">=", min_bin])
+        conditions.append([histogram_alias, ">=", min_bin])
     if max_value is not None:
         max_bin = histogram_params.start_offset + histogram_params.bucket_size * num_buckets
-        conditions.append([get_function_alias(histogram_col), "<=", max_bin])
+        conditions.append([histogram_alias, "<=", max_bin])
 
+    columns = [] if key_column is None else [key_column]
     results = query(
-        selected_columns=[key_col, histogram_col, "count()"],
+        selected_columns=columns + [histogram_column, "count()"],
         conditions=conditions,
         query=user_query,
         params=params,
-        # the zerofill step assumes it is ordered by the bin then name
-        orderby=[get_function_alias(histogram_col), get_function_alias(key_col)],
-        limit=len(measurements) * num_buckets,
+        orderby=[histogram_alias],
+        limit=len(fields) * num_buckets,
         referrer=referrer,
         auto_fields=True,
         use_aggregate_conditions=True,
         functions_acl=["array_join", "histogram"],
     )
 
-    return normalize_measurements_histogram(
-        measurements, num_buckets, key_col, histogram_params, results
-    )
+    return normalize_histogram_results(fields, key_column, histogram_params, results)
+
 
+def get_histogram_column(fields, key_column, histogram_params):
+    """
+    Generate the histogram column string.
+
+    :param [str] fields: The list of fields for which you want to generate the histograms for.
+    :param str key_column: The column for the key name. This is only set when generating a
+        multihistogram of measurement values. Otherwise, it should be `None`.
+    :param HistogramParms histogram_params: The histogram parameters used.
+    """
 
-def get_measurements_histogram_col(params):
-    return u"histogram(measurements_value, {:d}, {:d}, {:d})".format(
-        params.bucket_size, params.start_offset, params.multiplier
+    field = fields[0] if key_column is None else "measurements_value"
+    return u"histogram({}, {:d}, {:d}, {:d})".format(
```

Comment:
Could this format cause a ValueError if a string get this far in `bucket_size`? Can a string get this far?

Reply:
Non-integer values of `numBuckets`, `precision`, `min`, `max` are all guarded against in the `HistogramSerializer`. See https://github.com/getsentry/sentry/blob/4863d51274f26b14edfca0f53e835e488289f100/src/sentry/api/endpoints/organization_events_histogram.py#L22-L25. The parameters here are all derived from them and guaranteed to be integers.

Additionally, there are tests for this as well. Example https://github.com/getsentry/sentry/blob/4863d51274f26b14edfca0f53e835e488289f100/tests/snuba/api/endpoints/test_organization_events_histogram.py#L158-L168

Reply:
My mistake. I didn't look at the tests before making this comment :blush: